docs/rules/no-side-effects-wait-for.md

```diff
@@ -0,0 +1,73 @@
+# Side effects inside `waitFor` are not preferred (no-side-effects-wait-for)
+
+## Rule Details
+
+This rule aims to avoid the usage of side effects actions (`fireEvent` or `useEvent`) inside `waitFor`.
+Since `waitFor` is intended for things that have a non-deterministic amount of time between the action you performed and the assertion passing,
+the callback can be called (or checked for errors) a non-deterministic number of times and frequency.
+This will make your side-effect run multiple times.
+
+Example of **incorrect** code for this rule:
+
+```js
+const foo = async () => {
+  await waitFor(() => {
+    fireEvent.keyDown(input, { key: 'ArrowDown' });
+    expect(b).toEqual('b');
+  });
+
+  // or
+  await waitFor(function() {
+    fireEvent.keyDown(input, { key: 'ArrowDown' });
+    expect(b).toEqual('b');
+  });
+
+  // or
+  await waitFor(() => {
+    userEvent.keyDown(input, { key: 'ArrowDown' });
+    expect(b).toEqual('b');
+  });
+
+  // or
+  await waitFor(function() {
+    userEvent.keyDown(input, { key: 'ArrowDown' });
+    expect(b).toEqual('b');
+  });
+};
+```
+
+Examples of **correct** code for this rule:
+
+```js
+const foo = async () => {
+  fireEvent.keyDown(input, { key: 'ArrowDown' });
+  await waitFor(() => {
+    expect(b).toEqual('b');
+  });
+
+  // or
+  fireEvent.keyDown(input, { key: 'ArrowDown' });
+  await waitFor(function() {
+    expect(b).toEqual('b');
+  });
+
+  // or
+  userEvent.keyDown(input, { key: 'ArrowDown' });
+  await waitFor(() => {
+    expect(b).toEqual('b');
+  });
+
+  // or
+  userEvent.keyDown(input, { key: 'ArrowDown' });
+  await waitFor(function() {
+    expect(b).toEqual('b');
+  });
+};
+```
```

Comment:
I am a bit confused about the 
```ts
const foo = async () => {
```

parts. I would just remove them and leave the examples directly.

perhaps I would also add a small comment explaining what the snippet does (like `// using arrow function`) or similar to be more consistent with other rules (not all, but many do this)

Reply:
Ops! This is something extra here 😅